### DISCUSSION TITLE
graphviz for clang*

### DISCUSSION THREAD

**Discussion by revelator** [msys2/MINGW-packages#9253] [General]
[graphviz-packages.zip](https://github.com/msys2/MINGW-packages/files/6910926/graphviz-packages.zip)

bare bones graphviz packages for clang*.
one small bug remains which means clangs windres cannot be used for building them yet, lefty seems to have a bug in it's resource file causing clangs windres to throw an invalid language id error, it builds cleanly with clang if i use windres from binutils but also means it will be indigestable for a ci build. You can use these untill i find a workaround for the language id bug.

**Comment by Biswa96** [comment]
> You can use these untill i find a workaround for the language id bug

The language id can be passed as integer in base 10 (i.e. 1033) instead of hex. I have tried it. But graphviz failed to build with ucrt, need to relocate path. I have been talking with upstream recently.

**Reply by revelator** [reply]
besides the language id bug i had no problems building it against ucrt with the latest clang :), seems to work also as i just built vala with graphviz support.

**Reply by Biswa96** [reply]
I mean UCRT64 msystem, not the CLANG64 msystem. autoconf issue.

**Reply by revelator** [reply]
ah ok well these are for the clang msystem, but since both use ucrt i wonder ? autoconf having trouble detecting the correct build system for UCRT64 maybe.

**Reply by jeremyd2019** [reply]
try to make a minimal reproducer rc file that works with windres (and MS rc.exe) but fails on llvm-windres, and open an issue and tag mstorsjo, he's been fixing some rc file issues like that recently.

**Reply by Biswa96** [reply]
😁 
* `windres --language 0x1 -i /dev/null -o /dev/null` fails
* `windres --language 1 -i /dev/null -o /dev/null` works

**Reply by jeremyd2019** [reply]
oh, so it's command-line parsing not even rc parsing

**Reply by revelator** [reply]
would seem so :)

**Reply by revelator** [reply]
in graphviz case it is 0x409

**Reply by jeremyd2019** [reply]
https://bugs.llvm.org/show_bug.cgi?id=51295

**Reply by jeremyd2019** [reply]
https://reviews.llvm.org/D107263.  It turns out the language id is always interpreted as hex even without a leading 0x, so it would be `-l 409` not `-l 1033`

**Reply by revelator** [reply]
ah ok i though it was because it only accepted decimal, well easy fix untill they get it fixed upstream :)

**Comment by revelator** [comment]
so should be ->

```
if RC_IS_RC
lefty.res : ws/mswin32/lefty.rc
	rc -l 1003 -Iws/mswin32 '-IC:\\progra~1\\microsoft visual studio 8\\vc\\PlatformSDK\\include' -DNDEBUG -fo lefty.res ws/mswin32/lefty.rc
else
lefty.res : ws/mswin32/lefty.rc
	$(RC) -l 1003 -I$(top_srcdir)/cmd/lefty/ws/mswin32 -DNDEBUG -Ocoff -o lefty.res $(top_srcdir)/cmd/lefty/ws/mswin32/lefty.rc
endif
```

**Comment by revelator** [comment]
[mingw-w64-graphviz.zip](https://github.com/msys2/MINGW-packages/files/6912085/mingw-w64-graphviz.zip)

fixed windres error so should work with clang now :)

**Comment by jeremyd2019** [comment]
I tried to use this to test the fix to llvm-windres, but it depends on libglade, which is not yet built because that makedepends on python2.  It only needs python for a single script `libglade-convert`, and it looks like it would mostly work on python3.  I'm inclined to add a call to 2to3 to the PKGBUILD and switching the dependency over to python.

**Reply by revelator** [reply]
hmm cannot remember precisely what i did but libglade works just fine with python3 :)

**Reply by jeremyd2019** [reply]
yeah it's just this one script, and it builds fine with python3, or indeed with no python at all, just the script won't work if something tries to use it.

**Comment by revelator** [comment]
here you go i used this patch from opensuse leap :) 
[libglade2-python3.zip](https://github.com/msys2/MINGW-packages/files/6947359/libglade2-python3.zip)

**Reply by jeremyd2019** [reply]
yeah those are the same changes that 2to3 makes :wink:

**Reply by revelator** [reply]
hehe ;)

**Reply by revelator** [reply]
btw need to build with -j1 or clang will bork because of a race condition.

**Reply by revelator** [reply]
one small problem i noticed is that libglade-convert has indentation errors which python2 accepted but python3 does not.
luckily it seems it does not affect building the few things that rely on it :)

**Reply by jeremyd2019** [reply]
saw that - #9304

**Comment by revelator** [comment]
just need to point the python in PKGBUILD to python3 instead of python2 and things will work.

**Comment by jeremyd2019** [comment]
Added my version of graphviz fixes to #9304 since they depend on libglade, so CI can build both.

**Reply by jeremyd2019** [reply]
forgot that graphviz depends on the updated clang that isn't merged yet. 🤦

**Comment by jeremyd2019** [comment]
It looks like graphviz is embedding absolute paths on the build machine via `-D` C preprocessor defines.  This is causing an error on UCRT because they were not properly escaping the backslashes and `\ucrt` isn't a valid unicode escape.  But the bigger problem is that these paths will not necessarily be the same on user installs.

**Reply by jeremyd2019** [reply]
may actually be meant to be a relative path... it starts it with `.$(PATH_SEPARATOR)`.
```
-DDFLT_GVPRPATH="\".$(PATH_SEPARATOR)$(pkgdatadir)/gvpr\""
```

I don't see how that works in *nix either, that would default to ./usr/share/something

**Reply by jeremyd2019** [reply]
Oh, PATH_SEPARATOR, not DIRECTORY_SEPARATOR, so more like `.:/usr/share/something`.  OK, that makes more sense

**Reply by revelator** [reply]
aye same path delimiter the old msys shell used it seems :)

**Reply by revelator** [reply]
makes sense since the fist graphviz mingw port was actually done on msys/mingw i remember it was hell to build though xD

**Reply by jeremyd2019** [reply]
Well I'm planning to just swap the backslashes with forward slashes (I started out trying to escape the backslashes, but I couldn't find the proper number of backslashes to make it through).  If somebody cares enough, this would be a good place to pull in pathtools and get the path relative to the exe.

**Reply by revelator** [reply]
that should work :)

**Reply by jeremyd2019** [reply]
argh, it's apparently the msys command line conversion stuff.  the path going into the compiler is unixy.  I'm going to punt it for now and just remove the define from the makefile.  In the .c file if it's not defined it's defined to `"."` (rather than `".;D:\a\_temp\msys\msys64\clang64\share\graphviz\gvpr"` which won't be valid anywhere other than the builder anyway)

**Reply by revelator** [reply]
hmm seems having imagemagick breaks the swig plugins, if enabled it puts a load of garbadge into the PYTHON2 variable of all things Oo

**Comment by jeremyd2019** [comment]
#9305
